xrpl/clients/websocket_client.py

```diff
@@ -0,0 +1,150 @@
+
+        Returns:
+            The response from the server, as a Response object.
+        """
+        return asyncio.get_event_loop().run_until_complete(
+            self.request_async(request_object)
+        )
+
+    async def listen_async(
+        self: WebsocketClient,
+        request_object: Request,
+        handler: Callable[[Response], None],
+    ) -> None:
+        """
+        Asynchronously submits a request represented by the request_object to the
+        rippled node specified by this client's URL, and listens to (and processes) all
+        responses from the server.
+
+        Arguments:
+            request_object: An object representing information about a rippled request.
+            handler: The method that handles the Response objects from the server.
+                Takes a Response object as a parameter and returns None.
+        """
+        formatted_request = request_to_websocket(request_object)
+        if "id" not in formatted_request:
+            formatted_request["id"] = "request_{}".format(formatted_request["command"])
+
+        async with websockets.connect(self.url) as websocket:
+            self.websockets.add(websocket)
+            await websocket.send(json.dumps(formatted_request))
+            try:
+                async for message in websocket:
+                    await websocket.send(json.dumps(formatted_request))
+                    response = await websocket.recv()
+                    response_dict = json.loads(response)
+                    _check_ids(formatted_request, response_dict)
+                    handler(websocket_to_response(response_dict))
+            except websockets.exceptions.ConnectionClosedOK:
+                return
+            finally:
+                if websocket in self.websockets:
+                    # could have been removed by self.close_async()
+                    self.websockets.remove(websocket)
+
+    def listen(
+        self: WebsocketClient,
+        request_object: Request,
+        handler: Callable[[Response], None],
+    ) -> None:
+        """
+        Synchronously submits a request represented by the request_object to the
+        rippled node specified by this client's URL, and listens to (and processes) all
+        responses from the server.
+
+        Arguments:
+            request_object: An object representing information about a rippled request.
+            handler: The method that handles the Response objects from the server.
+                Takes a Response object as a parameter and returns None.
+        """
+        asyncio.get_event_loop().run_until_complete(
+            self.listen_async(request_object, handler)
+        )
+
+    async def close_async(self: WebsocketClient) -> None:
+        """Asynchronously closes any open WebSocket connections."""
+        while len(self.websockets) != 0:
+            websocket = self.websockets.pop()
+            await websocket.close()
```

Comment:
worried that this is waiting for each connection to close before requesting the next closure, which is a performance miss. also, confused by this approach since we should have a single socket connection, not multiple